agent/agent.go

```diff
@@ -0,0 +1,235 @@
+	// client. (Optional)
+	BalanceCallback func(store.Balance)
+
+	// PoolMessageCallback is called whenever the client receives a message
+	// from the pool. This can be a welcome message including rules and
+	// instructions for how to manage the client's balance. It should be
+	// displayed to the client. (Optional)
+	PoolMessageCallback func(string)
+
+	// NumHosts is the minimum number of vipnode hosts the client should
+	// maintain connections with. (Optional)
+	NumHosts int
+
+	// Payout is the address to register to associate pool credits towards.
+	// (Optional)
+	Payout string
+
+	mu       sync.Mutex
+	started  bool
+	stopCh   chan struct{}
+	waitCh   chan error
+	nodeInfo ethnode.UserAgent // cached during Start
+}
+
+// Start registers the node on the given pool and starts sending peer updates
+// every store.KeepaliveInterval. It returns after
+// successfully registering with the pool.
+func (a *Agent) Start(p pool.Pool) error {
+	a.mu.Lock()
+	if a.started {
+		a.mu.Unlock()
+		return errors.New("pool already started")
+	}
+	if a.stopCh == nil {
+		a.stopCh = make(chan struct{})
+		a.waitCh = make(chan error, 1)
+	}
+	a.mu.Unlock()
+
+	startCtx, cancel := context.WithTimeout(context.Background(), startTimeout)
+	defer cancel()
+
+	enode, err := a.EthNode.Enode(startCtx)
+	if err != nil {
+		return err
+	}
+	logger.Printf("Connected to local node: %s", enode)
+
+	version := a.Version
+	if version == "" {
+		version = "dev"
+	}
+
+	connectReq := pool.ConnectRequest{
+		Payout:         a.Payout,
+		NodeURI:        a.NodeURI,
+		VipnodeVersion: version,
+		NodeInfo:       a.EthNode.UserAgent(),
+	}
+	a.nodeInfo = connectReq.NodeInfo
+	resp, err := p.Connect(startCtx, connectReq)
+	if err != nil {
+		return err
+	}
+	logger.Printf("Registered on pool: Version %s", resp.PoolVersion)
+
+	if resp.Message != "" && a.PoolMessageCallback != nil {
+		a.PoolMessageCallback(resp.Message)
+	}
+
+	if err := a.updatePeers(startCtx, p); err != nil {
+		return err
+	}
+
+	go func() {
+		a.waitCh <- a.serveUpdates(p)
+	}()
+	return nil
+}
+
+// Whitelist a peer for this node.
+func (a *Agent) Whitelist(ctx context.Context, nodeID string) error {
+	logger.Printf("Received whitelist request: %s", nodeID)
+	return a.EthNode.AddTrustedPeer(ctx, nodeID)
+}
+
+// Stop shuts down all the active connections cleanly.
+func (a *Agent) Stop() {
+	a.stopCh <- struct{}{}
+}
+
+// Wait blocks until the agent is stopped. It returns any errors that occur
+// during stopping.
+func (a *Agent) Wait() error {
+	return <-a.waitCh
+}
+
+func (a *Agent) serveUpdates(p pool.Pool) error {
+	ticker := time.Tick(store.KeepaliveInterval)
+	for {
+		select {
+		case <-ticker:
+			// FIXME: Sametimes a.updatePeers could take a while, does it make
+			// sense to keep ticking every interval regardless? Or should we
+			// make sure there is an offset between calls? What if there's
+			// overlap, do we want a mutex?
```

Comment:
So the `Ticker` docs mention:

> adjusts the intervals or drops ticks to make up for slow receivers

But based on the discussion [here](https://github.com/golang/go/issues/6142) that just applies to if an entire tick is missed.  

There shouldn't be any overlap since AFAICT updatePeers is synchronous, but it is possible that two calls stack up pretty close to one another.  But IMO that's acceptable trade-off to keeping the code pretty straight forward.

Reply:
TIL! I think that'll do the trick, for now at least. There might be a scenario where if we have toooons of nodes, we might want to be better about staggering their updates but I don't think we'll hit that anytime soon.